### PR TITLE
WIP: Move DateTime native code into managed

### PR DESCRIPTION
**Pull request by filipnavara**
Just testing build and unit tests now. If it works out then the code could be shared between CoreCLR and CoreRT completely.

### CODE REVIEW

**Comment by jkotas** [comment]
Some this is implemented as FCalls to avoid PInvoke overhead. This would need performance testing as well.

**Comment by filipnavara** [comment]
@dotnet-bot test Windows_NT arm Cross Debug Innerloop Build
@dotnet-bot test Windows_NT arm64 Cross Debug Innerloop Build
@dotnet-bot test Windows_NT x64 Release CoreFX Tests

**Comment by filipnavara** [comment]
I've done some preliminary benchmarks and it shows some rather interesting things. Firstly, my initial implementation with `LazyInitializer` and delegates is about 2x slower than the baseline. Switching to an approach with statically initialized `s_systemSupportsPreciseSystemTime` variable and simple `if` condition cuts down the difference quite a lot:

Method | Mean | Error | StdDev
-- | -- | -- | --
UtcNow [Old] | 25.32 ns | 0.4442 ns | 0.3938 ns
UtcNow [New] | 30.75 ns | 0.6206 ns | 0.5805 ns

Now, these numbers are not completely representative since it was run on .NET Core 2.1.6 runtime with a code copied to a separate BenchmarkDotNet project. Perhaps there's a better way to benchmark something like this?

Another interesting observation is that the `GetSystemTimePreciseAsFileTime` is nearly twice as slow on my system as `GetSystemTimeAsFileTime`.

**Comment by Suchiman** [comment]
@filipnavara 
> Another interesting observation is that the GetSystemTimePreciseAsFileTime is nearly twice as slow on my system as GetSystemTimeAsFileTime.

That seems to be expected: https://github.com/dotnet/coreclr/pull/9736

File: src/System.Private.CoreLib/src/System/DateTime.Windows.cs [hunk 1]
```diff
@@ -4,24 +4,40 @@
...
 
 namespace System
 {
     public readonly partial struct DateTime
     {
+        private unsafe delegate void GetSystemTimeAsFileTimeDelegate(long* lpSystemTimeAsFileTime);
+
         internal static readonly bool s_systemSupportsLeapSeconds = SystemSupportsLeapSeconds();
+        internal static readonly bool s_systemSupportsPreciseSystemTime = SystemSupportsPreciseSystemTime();
+        private static GetSystemTimeAsFileTimeDelegate _getSystemTimeAsFileTime;
```
Comment (marek-safar) [inline src/System.Private.CoreLib/src/System/DateTime.Windows.cs]: Where is this field used/set?

Comment (filipnavara) [reply src/System.Private.CoreLib/src/System/DateTime.Windows.cs]: It is not. I forgot to delete it.

**Comment by filipnavara** [comment]
I had some flaws in my benchmark. This is current state with .NET Core 3.0 Preview 1 runtime:

Method | Mean | Error | StdDev
-- | -- | -- | --
UtcNow [Old] | 25.34 ns | 0.1404 ns | 0.1313 ns
UtcNow [New] | 27.81 ns | 0.2964 ns | 0.2628 ns

Is this worth pursuing further? Maybe there are ways to squeeze the last few nanoseconds back.

**Comment by danmosemsft** [comment]
@filipnavara UtcNow is often on the hot path, yes I think it is worth investigating further perf improvements.

**Comment by filipnavara** [comment]
Well, I can scale back the PR and keep the FCalls and move just the rest to shared partition. I am not sure if I will be able to squeeze any more performance out of the P/Invoke approach.

**Comment by jkotas** [comment]
The performance is closer than what I thought it is going to be. I will take a look at this in detail when I get a chance.

**Comment by filipnavara** [comment]
Thanks! Any advice or help is appreciated.

**Comment by filipnavara** [comment]
There are some particularly bad things in the code gen, which could likely be eliminated:
```
00007ff8`c8b1f667 488945b0        mov     qword ptr [rbp-50h],rax
00007ff8`c8b1f66b eb00            jmp     00007ff8`c8b1f66d // Jumping to next instruction, really?
00007ff8`c8b1f66d 488b45b0        mov     rax,qword ptr [rbp-50h] // Loading whatever we wrote there two instructions ago
```

**Comment by mikedn** [comment]
> There are some particularly bad things in the code gen, which could likely be eliminated:

Hmm, can you post more code? I've seen such jumps before but they're not common.

**Comment by filipnavara** [comment]
@mikedn I extracted the PInvoke alone into a small function

```c#
        public unsafe void JustPInvoke()
        {
            long timestamp;
            Interop.Kernel32.GetSystemTimePreciseAsFileTime(&timestamp);
        }
```
and looked at the generated code (.NET Core 3.0.0-preview-27122-01 (CoreCLR 4.6.27121.03, CoreFX 4.7.18.57103), 64bit RyuJIT):

```assembly
; DateTimeTest.DateTimeBenchmark.JustPInvoke()
       push    rbp
       push    r15
       push    r14
       push    r13
       push    r12
       push    rdi
       push    rsi
       push    rbx
       sub     rsp,78h
       lea     rbp,[rsp+0B0h]
       xor     edx,edx
       mov     qword ptr [rbp-40h],rdx
       mov     qword ptr [rbp-48h],rdx
       mov     qword ptr [rbp+10h],rcx
       lea     rcx,[rbp-80h]
       mov     rdx,r10
       call    coreclr+0xad090
       mov     qword ptr [rbp-48h],rax
       mov     rcx,rsp
       mov     qword ptr [rbp-60h],rcx
       mov     rcx,rbp
       mov     qword ptr [rbp-50h],rcx
       lea     rcx,[rbp-40h]
       xor     r11d,r11d
       mov     rax,7FF89F1F39D8h
       mov     qword ptr [rbp-70h],rax
       lea     rax,[00007ff8`9f0affde]
       mov     qword ptr [rbp-58h],rax
       mov     rax,qword ptr [rbp-48h]
       lea     rdx,[rbp-80h]
       mov     qword ptr [rax+10h],rdx
       mov     rax,qword ptr [rbp-48h] ; This is loading a value that was already loaded 4 lines above
       mov     byte ptr [rax+0Ch],0
       call    qword ptr [00007ff8`9f1f3b70]
       mov     rdx,qword ptr [rbp-48h]
       mov     byte ptr [rdx+0Ch],1
       cmp     dword ptr [coreclr!g_CLREngineMetrics+0x9080
       je      M00_L00
       call    qword ptr [coreclr!CreateCLRProfiling+0x20b188
M00_L00:
       mov     rax,qword ptr [rbp-48h]
       mov     rdx,qword ptr [rbp-78h]
       mov     qword ptr [rax+10h],rdx
       jmp     M00_L01 ; Unnecessary jump
M00_L01:
       mov     rax,qword ptr [rbp-48h] ; Unnecessary load (probably not eliminated because of the jump?)
       mov     byte ptr [rax+0Ch],1
       lea     rsp,[rbp-38h]
       pop     rbx
       pop     rsi
       pop     rdi
       pop     r12
       pop     r13
       pop     r14
       pop     r15
       pop     rbp
       ret
...
; Total bytes of code 204
```

It could very well be something screwed up by BenchmarkDotNet or my .NET Core installation. Smallest repro code here:

```c#
using System;
using System.Runtime.InteropServices;

public class C {
        [DllImport("kernel32", ExactSpelling = true, SetLastError = false)]
        internal static unsafe extern void GetSystemTimePreciseAsFileTime(long* lpSystemTimeAsFileTime);

    	public unsafe void JustPInvoke()
        {
            long timestamp;
            GetSystemTimePreciseAsFileTime(&timestamp);
        }
}
```

**Comment by filipnavara** [comment]
@AndyAyersMS Yeah, there's BDN bug filled for that (https://github.com/dotnet/BenchmarkDotNet/issues/899). I was seeing some weird output with `COMPlus_TieredCompilation=0` earlier, but likely due to some other mistake I made. Your output is way more readable than whatever I managed to get locally.

**Comment by Suchiman** [comment]
@filipnavara you can get that Output via https://github.com/dotnet/coreclr/blob/master/Documentation/building/viewing-jit-dumps.md

**Comment by stephentoub** [comment]
> Another interesting observation is that GetSystemTimePreciseAsFileTime is nearly twice as slow

Yeah, we decided in https://github.com/dotnet/coreclr/pull/9736 that the performance difference was worth it for the accuracy.

**Comment by filipnavara** [comment]
Closing for now, thanks for all the valuable feedback. The Unix part was superseded by #22383 and the P/Invoke performance improvements are discussed there and in #22320.

**Comment by filipnavara** [comment]
Note to myself: This should be resurrected now that `SuppressGCTransition` attribute is available.

**Pull request closed without merging**